File: iosApp/deadly/Core/Database/DAOs/ShowPlayerTagDAO.swift
import GRDB

struct ShowPlayerTagDAO: Sendable {
    let database: AppDatabase

    // MARK: - Upsert

    func upsert(_ record: ShowPlayerTagRecord) throws {
        try database.write { db in
            var r = record
            try r.save(db, onConflict: .replace)
        }
    }

    // MARK: - Fetch

    func fetchForShow(_ showId: String) throws -> [ShowPlayerTagRecord] {
        try database.read { db in
            try ShowPlayerTagRecord
                .filter(Column("showId") == showId)
                .order(Column("playerName").asc)
                .fetchAll(db)
        }
    }

    func fetchStandoutShowsForPlayer(_ playerName: String) throws -> [ShowPlayerTagRecord] {
        try database.read { db in
            try ShowPlayerTagRecord
                .filter(Column("playerName") == playerName && Column("isStandout") == true)
                .order(Column("createdAt").desc)
                .fetchAll(db)
        }
    }

    func fetchCountForShow(_ showId: String) throws -> Int {
        try database.read { db in
            try ShowPlayerTagRecord.filter(Column("showId") == showId).fetchCount(db)
        }
    }

    func fetchAll() throws -> [ShowPlayerTagRecord] {
        try database.read { db in
            try ShowPlayerTagRecord.fetchAll(db)
        }
    }

    func fetchAllStandoutPlayerNames() throws -> [String] {
        try database.read { db in
            try String.fetchAll(db, sql: """
                SELECT DISTINCT playerName FROM show_player_tags WHERE isStandout = 1 ORDER BY playerName ASC
            """)
        }
    }

    // MARK: - Delete

    func remove(showId: String, playerName: String) throws {
        try database.write { db in
            try db.execute(
                sql: "DELETE FROM show_player_tags WHERE showId = ? AND playerName = ?",
                arguments: [showId, playerName]
            )
        }
    }

    func removeForShow(_ showId: String) throws {
        try database.write { db in
            try ShowPlayerTagRecord.filter(Column("showId") == showId).deleteAll(db)
        }
    }
}

File: androidApp/core/database/src/main/java/com/grateful/deadly/core/database/dao/ShowPlayerTagDao.kt
package com.grateful.deadly.core.database.dao

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import com.grateful.deadly.core.database.entities.ShowPlayerTagEntity
import kotlinx.coroutines.flow.Flow

@Dao
interface ShowPlayerTagDao {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun upsert(tag: ShowPlayerTagEntity)

    @Query("SELECT * FROM show_player_tags WHERE showId = :showId ORDER BY playerName ASC")
    suspend fun getTagsForShow(showId: String): List<ShowPlayerTagEntity>

    @Query("SELECT * FROM show_player_tags WHERE showId = :showId ORDER BY playerName ASC")
    fun getTagsForShowFlow(showId: String): Flow<List<ShowPlayerTagEntity>>

    @Query("SELECT * FROM show_player_tags WHERE playerName = :playerName AND isStandout = 1 ORDER BY createdAt DESC")
    suspend fun getStandoutShowsForPlayer(playerName: String): List<ShowPlayerTagEntity>

    @Query("DELETE FROM show_player_tags WHERE showId = :showId AND playerName = :playerName")
    suspend fun removeTag(showId: String, playerName: String)

    @Query("DELETE FROM show_player_tags WHERE showId = :showId")
    suspend fun removeTagsForShow(showId: String)

    @Query("SELECT COUNT(*) FROM show_player_tags WHERE showId = :showId")
    suspend fun getTagCountForShow(showId: String): Int

    @Query("SELECT DISTINCT playerName FROM show_player_tags WHERE isStandout = 1 ORDER BY playerName ASC")
    suspend fun getAllStandoutPlayerNames(): List<String>

    @Query("SELECT * FROM show_player_tags")
    suspend fun getAll(): List<ShowPlayerTagEntity>
}
